messaging/src/main/java/org/axonframework/messaging/Context.java

```diff
@@ -0,0 +1,179 @@
+     *
+     * @param key The key of the resource to check.
+     * @return {@code true} if a resource is registered with this {@code key}, otherwise {@code false}.
+     */
+    boolean containsResource(@Nonnull ResourceKey<?> key);
+
+    /**
+     * Returns the resource currently registered under the given {@code key}, or {@code null} if no resource is
+     * present.
+     *
+     * @param key The key to retrieve the resource for.
+     * @param <T> The type of resource registered under the given {@code key}.
+     * @return The resource currently registered under given {@code key}, or {@code null} if not present.
+     */
+    <T> T getResource(@Nonnull ResourceKey<T> key);
+
+    /**
+     * Constructs a copy of {@code this} {@link Context} adding the given {@code resources} under the given {@code key}
+     * to the copy.
+     *
+     * @param key      The key under which to register the {@code resource}.
+     * @param resource The resource to register.
+     * @param <T>      The type of resource registered under the given {@code key}.
+     * @return A copy of {@code this} {@link Context} with the added given {@code resources} under the given {@code key}
+     * to the copy.
+     */
+    <T> Context withResource(@Nonnull ResourceKey<T> key,
+                             @Nonnull T resource);
+
+    /**
+     * Object that is used as a key to retrieve and register resources of a given type in a {@link Context}.
+     * <p>
+     * Instance of a {@code ResourceKey} can be created using either {@link #create()}, {@link #create(Class)}, or
+     * {@link #create(String)}. The former option will construct a unique key at all times, while the {@link Class} and
+     * {@link String} based factory methods result in an identical {@code ResourceKey} if the {@code Class} or
+     * {@code String} is reused.
+     *
+     * @param <T> The type of resource registered under this key.
+     * @author Allard Buijze
+     * @author Mitchell Herrijgers
+     * @author Steven van Beelen
+     * @since 5.0.0
+     */
+    @SuppressWarnings("unused") // Suppresses the warning that the generic type is not used.
+    final class ResourceKey<T> {
+
+        private final Object identity;
+        private final String debugString;
+
+        private ResourceKey(@Nonnull Object identity,
+                            @Nonnull String debugString) {
+            this.identity = identity;
+            this.debugString = debugString;
+        }
+
+        /**
+         * Creates a {@link ResourceKey} using a {@link UUID#randomUUID() random UUID} as the key's identity.
+         *
+         * @param <T> The type of resource of this key.
+         * @return A {@link ResourceKey} using a {@link UUID#randomUUID() random UUID} as the key's identity, used for
+         * adding and retrieving context-specific resources.
+         */
+        public static <T> ResourceKey<T> create() {
```

Comment:
While this works, I would prefer to be explicit of the behavior of the create on the method name alone. 
Maybe we should rename this method to `createUnique()`? 
And, optionally, we can then name the other `createShared(String identity)`.

Reply:
Aaah nice, I like this specification. Otherwise it's only apparent from the JavaDoc, while the rename makes it clear from the get-go.
I did choose slightly different names too: `uniqueKey` and `sharedKey`. WDYT?